ArduCopter/mode_auto.cpp

```diff
@@ -1165,21 +1167,192 @@ void ModeAuto::payload_place_run()
+            // decent rate has not reached g.land_speed
+            nav_payload_place.descent_established_timestamp = now_ms;
+            break;
+        } else if (now_ms - nav_payload_place.descent_established_timestamp < descend_throttle_calibrate_time_ms) {
+            // record minimum thrust for descend_throttle_calibrate_time_ms
+            nav_payload_place.descend_throttle_level = MIN(nav_payload_place.descend_throttle_level, current_throttle_level);
+            nav_payload_place.place_start_timestamp = now_ms;
+            break;
+        } else if (current_throttle_level > g2.pldp_throttle_placed_fraction * nav_payload_place.descend_throttle_level) {
+            // thrust is above minimum threshold
+            nav_payload_place.place_start_timestamp = now_ms;
+            break;
+        } else if (is_positive(g2.pldp_range_finder_minimum)) {
+            if (!copter.rangefinder_state.enabled) {
+                // abort payload place because rangefinder is not enabled
+                nav_payload_place.state = PayloadPlaceStateType_Ascent_Start;
+                gcs().send_text(MAV_SEVERITY_WARNING, "%s PLDP_RNG_MIN set and rangefinder not enabled", prefix_str);
+                break;
+            } else if (copter.rangefinder_alt_ok() && (copter.rangefinder_state.glitch_count == 0) && (copter.rangefinder_state.alt_cm > g2.pldp_range_finder_minimum * 100.0)) {
+                // range finder altitude is above minimum
+                nav_payload_place.place_start_timestamp = now_ms;
+                break;
+            }
+        }
+
+        // If we get here:
+        // 1. we have reached decent velocity
+        // 2. measured the thrust level required for decent
+        // 3. detected that our thrust requirements have reduced
+        // 4. rangefinder range has dropped below minimum if set
+        // 5. place_start_timestamp has been initialised
+
+        // payload touchdown must be detected for 0.5 seconds
+
+        if (now_ms - nav_payload_place.place_start_timestamp > placed_time_ms) {
+            nav_payload_place.state = PayloadPlaceStateType_Release;
+            gcs().send_text(MAV_SEVERITY_INFO, "%s payload release thrust threshold: %f", prefix_str, static_cast<double>(g2.pldp_throttle_placed_fraction * nav_payload_place.descend_throttle_level));
+        }
+        break;
+
+    case PayloadPlaceStateType_Release:
+        // Reinitialise vertical position controller to remove discontinuity due to touch down of payload
+        pos_control->init_z_controller_no_descent();
+#if AP_GRIPPER_ENABLED == ENABLED
+        if (g2.gripper.valid()) {
+            gcs().send_text(MAV_SEVERITY_INFO, "%s Releasing the gripper", prefix_str);
+            g2.gripper.release();
+            nav_payload_place.state = PayloadPlaceStateType_Releasing;
+        } else {
+            gcs().send_text(MAV_SEVERITY_INFO, "%s Gripper not valid", prefix_str);
```

Comment:
In my S&R competition we didn't use a gripper but instead used a winch with a built in claw that automatically releases when the weight come off.  In this case I guess we would always see, "PayloadPlace: Gripper not valid".  It's not a deal breaker but it is a funny message to see when you know you're not using a gripper.

Reply:
Perhaps we should go to the delay state rather than the ascent state too.